File: docusaurus/docs/embedded/DefineEmbeddedObjects.m
// Define the embedded object
@interface Address : RLMEmbeddedObject
@property NSString *street;
@property NSString *city;
@property NSString *country;
@property NSString *postalCode;
@end

// Enable use in RLMArray
RLM_ARRAY_TYPE(Address)

@implementation Address
@end

// A Contact has one embedded address
@interface Contact : RLMObject
@property RLMObjectId *_id;
@property NSString *name;

// Define a single embedded object
@property Address *address;
@end

@implementation Contact
+ (NSString *)primaryKey {
    return @"_id";
}

+ (NSArray<NSString *> *)requiredProperties {
    return @[@"_id", @"name"];
}

+ (NSDictionary *)defaultPropertyValues {
    return @{@"_id": [RLMObjectId objectId]};
}

+ (instancetype)contactWithName:(NSString *)name {
    Contact *instance = [[Contact alloc] init];
    if (instance) {
        instance.name = name; 
    }
    return instance; 
}
@end

// A Business has multiple embedded addresses
@interface Business : RLMObject
@property RLMObjectId *_id;
@property NSString *name;
// Define an array of embedded objects
@property RLMArray<Address *><Address> *addresses;
@end

@implementation Business
+ (NSString *)primaryKey {
    return @"_id";
}

+ (NSArray<NSString *> *)requiredProperties {
    return @[@"_id", @"name"];
}
@end

File: source/examples/EmbeddedObjects/DefineEmbeddedObjects.swift
// Define an embedded object
class Address: EmbeddedObject {
    @objc dynamic var street: String? = nil
    @objc dynamic var city: String? = nil
    @objc dynamic var country: String? = nil
    @objc dynamic var postalCode: String? = nil
}

// Define an object with one embedded object
class Contact: Object {
    @objc dynamic var _id = ObjectId()
    @objc dynamic var name = ""
    @objc dynamic var address = Address() // Embed a single object
    
    override static func primaryKey() -> String? {
        return "_id"
    }
    
    convenience init(name: String, address: Address) {
        self.init()
        self.name = name
        self.address = address
    }    
}

// Define an object with an array of embedded objects
class Business: Object {
    @objc dynamic var _id = ObjectId()
    @objc dynamic var name = ""
    @objc dynamic var addresses: [Address] = [] // Embed an array of objects
    
    override static func primaryKey() -> String? {
        return "_id"
    }
    
    convenience init(name: String, addresses: [Address]) {
        self.init()
        self.name = name
        self.addresses = addresses
    }
}
